Add functionality for time-variable abundances and radiative losses

jwreep opened the pull request:
New functionality introduced to allow the abundance factor to vary with time, and thus the radiative losses to vary with time.

jwreep committed:
Add functionality to calculate abundance factor

jwreep committed:
add search function for nearest value in array

jwreep committed:
add misc.cpp and misc.h

jwreep committed:
add radiative loss data

jwreep committed:
add function to read abundance tables and calculate new losses

jwreep committed:
lots of debugging to get it to run

jwreep committed:
fixed format and input of radiative loss files

jwreep commented:
@wtbarnes, if we're changing abundances, does this function need to be updated?  [He/H] should be constant, but Z_avg likely changes.

https://github.com/rice-solar-physics/ebtelPlusPlus/blob/4918ca9475fb5503d465a2e6a19fcd1c99fdb966/source/loop.cpp#L416-L421

jwreep committed:
add correction for draining and later upflow events

jwreep committed:
use original radiation for initial conditions

jwreep committed:
add include guard for misc.h

Commit:
change bool use_variable_abundances to string radiation input

jwreep commented:
I slightly reworked the input functionality so that time-fixed coronal and time-fixed photospheric abundances are now valid options.

wtbarnes commented:
If you rebase or merge in main, you should pick up the C++-14 changes to SConstruct.

jwreep committed:
Merge branch 'rice-solar-physics:main' into main

Commit:
add radiation to test configurations

Commit:
revert config file to original values

Commit:
use the power law by default in config

jwreep commented:
Ready for review!  

Since this is a breaking change, I think we should review carefully.   Would any additional tests be useful?

wtbarnes commented:
> @wtbarnes, if we're changing abundances, does this function need to be updated?
> 
> https://github.com/rice-solar-physics/ebtelPlusPlus/blob/4918ca9475fb5503d465a2e6a19fcd1c99fdb966/source/loop.cpp#L416-L421

Honestly, I'm not sure. I need to go back and remind myself what the logic of that function even is. I think it mainly has to do with the H/He ratio though. How impacted is that by the transition from coronal to photospheric?

wtbarnes commented on source/loop.cpp:
```diff
@@ -42,6 +42,24 @@ Loop::Loop(char *config)
+          (parameters.radiation == "photospheric") ||
+          (parameters.radiation == "coronal") )
+  {
+      parameters.use_variable_abundances = true;
```
It is confusing that this is set to true even when the radiation uses only photospheric or only coronal abundances.

wtbarnes replied:
Instead of `use_variable_abundances`, why not make it `use_power_law_losses` and then invert the logic?

jwreep replied:
How about renaming to `use_lookup_table_losses` or something?  I think that more accurately describes the usage of this `bool`.

wtbarnes commented on source/helper.h:
```diff
@@ -66,6 +75,15 @@ struct Parameters {
+  double log10_loss_rate_array[101][7][7];
+  double log10_temperature_array[101];
+  double initial_density;
+  double previous_density;
+  double initial_abundance_factor;
+  double previous_abundance_factor;
+  bool upflowing, initial_radiation;
```
Could you add brief comments explaining what each of these is? Some are self-explanatory but others (e.g. `bool upflowing`) are not

wtbarnes commented on source/helper.h:
```diff
@@ -66,6 +75,15 @@ struct Parameters {
   size_t N;
+  
+  /* Arrays for variable abundance radiative losses */
+  double log10_loss_rate_array[101][7][7];
+  double log10_temperature_array[101];
+  double initial_density;
```
Hardcoding the size of these arrays makes me nervous, especially since the size appears in three separate places in the code: here, the read function, and the radiative loss calculation. Could these be vectors instead so that their sizes don't have to be declared here?

wtbarnes commented on source/loop.cpp:
```diff
@@ -116,6 +139,11 @@ state_type Loop::CalculateInitialConditions(void)
 
+  if( parameters.use_variable_abundances )
+  {
+      parameters.initial_radiation = true;
```
Why is this set to true here and then later set to false in line 174?

jwreep replied:
The electron density isn't defined prior to the calculation of the initial conditions, but is required to use the look-up tables for radiative losses.  The call to `CalculateC1` on line 151 has a nested call to calculate the radiative losses. 

So, I've defined a bool to check whether a call to `CalculateC1` is this initial conditions calculation, and if so, it uses the power law loss rate.  

In short, the initial conditions always use the power law losses because we don't yet have the electron density to do otherwise.  

There might be a more elegant way to do this.

wtbarnes replied:
Ah I see. That makes sense. Could you just add a short comment that just says that?

wtbarnes commented on source/loop.cpp:
```diff
@@ -350,6 +409,64 @@ double Loop::CalculateRadiativeLoss(double temperature)
 
+double Loop::CalculateRadiativeLoss(double temperature, double density)
+{
+    double abundance_array[] = {1.0, 1.5, 2.0, 2.5, 3.0, 3.5, 4.0};  // Discrete values available for abundance factors
```
Hardcoding these values makes me really nervous. My preferred solution would be that they are stored in the file(s) and then read in from there. Otherwise, we run the risk of the files and these hardcoded values being out of sync. It is also makes it easier to add new abundance factors if we do it this way.

wtbarnes commented on source/loop.cpp:
```diff
@@ -350,6 +409,64 @@ double Loop::CalculateRadiativeLoss(double temperature)
+    // Find the nearest value of AF to the values in our discrete list, then return that value
+    int abundance_index = find_closest(abundance_factor, abundance_array, array_length);
+    
+    double log10_density_array[] = {8.0, 8.69897, 9.0, 9.69897, 10.0, 10.69897, 11.0}; // Discrete values available for density
```
Same comment as above. I really don't like hardcoding these values like this and would prefer storing the values of the density in the files themselves. That way, if the density array ever changes, we don't have to update it in multiple places.

wtbarnes commented on source/loop.cpp:
```diff
@@ -420,12 +546,60 @@ void Loop::CalculateAbundanceCorrection(double helium_to_hydrogen_ratio)
+   std::string filenames[] = {"abund_10_rad_loss.dat","abund_15_rad_loss.dat","abund_20_rad_loss.dat","abund_25_rad_loss.dat",
+                                "abund_30_rad_loss.dat","abund_35_rad_loss.dat","abund_40_rad_loss.dat"};
+   std::ifstream fin;
+   int n_files = sizeof(filenames)/sizeof(filenames[0]);
```
Is this really the only way to get the length a list/array in modern C++?

jwreep replied:
Sadly, I believe so.  For C-style arrays, this is the "traditional" way to find lengths.  You could also use a for loop to count the number of elements -- which is possibly faster for small arrays.

The alternative is to use `std::vector`, which then has a built-in function for length.

jwreep replied:
This whole bit of code can be done more elegantly, but requires C++ 17 or more recent.  We don't need to hardcode the filenames, and can instead read in the directory and determine the filenames and number of files.  

Any objection to requiring C++ 17?  I've pushed a commit to do so, but can revert if we'd prefer not.  There does appear to be `boost` functions that could be used in its place.

wtbarnes replied:
No objection from me!

wtbarnes commented on source/loop.cpp:
```diff
@@ -420,12 +546,60 @@ void Loop::CalculateAbundanceCorrection(double helium_to_hydrogen_ratio)
+   {
+       fin.open(path+filenames[i]);
+       
+       for (int j=0; j < 100; ++j)  // Loop over temperatures (rows in files)
+       {
+           
+           for (int k=0; k < 8; ++k)   // Loop over densities (columns in files)
```
As before, hardcoding these lengths makes me really nervous.

jwreep replied:
Let me see if there's a way to pull the array sizes directly from the files.  This is certainly easier in Python...

jwreep replied:
I've removed all the hard-coded values.  I need to validate the output, but that's enough for now.  It's at least running without NaNs, so it has that going for it, which is nice.

jwreep replied:
All the plots in the paper were reproduced with the update.

wtbarnes reviewed (changes requested):
Sorry this has taken so long to get to. Overall, I'm fine with the user-facing part of this functionality. However, I'm mainly concerned with hardcoding the sizes of the arrays and the corresponding abundance factors and densities these correspond to. I've left several comments to this effect.

I also think we should add a few tests that use these three new inputs just to ensure that they keep working. I'm happy to do that.

jwreep commented:
> Honestly, I'm not sure. I need to go back and remind myself what the logic of that function even is. I think it mainly has to do with the H/He ratio though. How impacted is that by the transition from coronal to photospheric?
> 
> I'm also a little worried about the name of this function given that it potentially is unrelated to the abundance variations in the radiative losses. I'm fine with changing it to something else.

It looks like that's what this function does.  Helium does not change its abundance in what we've programmed here, although I've been tinkering with the idea of generalizing this for other stellar classes ([some stars have an inverse FIP bias](https://ui.adsabs.harvard.edu/abs/2008A%26A...482..639N/abstract)).  In that case, it's possible that helium might change with time.

Commit:
add descriptions of variables in helper.h

Commit:
revert the config file

Commit:
rename function

Commit:
remove hardcoding of radiation filenames and number of files -- requires > C++ 17

Commit:
use_variable_abundances -> use_lookup_table_losses

Commit:
remove hardcoding of indices

Commit:
add comment about initial conditions radiation

jwreep commented:
@wtbarnes, I'm having trouble adding tests, and not sure what the issue is.  I keep getting something similar to this when running `pytest`, for even very simple tests:
```
========================================================= short test summary info =========================================================
ERROR tests/test_solver.py::test_coronal_output - FileNotFoundError: /var/folders/5r/lmtf450n5fn0xrgg3t8689m40000gn/T/tmpesgpj0th/ebtelplusplus.tmp not found.
```
The code I've added to `tests/test_solver.py`: 
```
@pytest.fixture(scope='module')
def coronal_results(base_config):
    config = base_config.copy()
    config['radiation'] = "coronal"
    return run_ebtelplusplus(config)

def test_coronal_output(coronal_results):
    time = coronal_results['time'].to_value('s')
    assert np.isclose(time[0], 0.0)
```

Any idea here?  

Other than the tests, I think I've addressed everything else.

wtbarnes commented on source/util/misc.cpp:
```diff
@@ -0,0 +1,45 @@
+#include "misc.h"
+
+
+int find_closest(double x, double array[], int array_length)
+{
+    /* Traverses the whole array, so O(N) search.  
+     * A binary search would be more efficient for long arrays: O(ln N). */
+    int index = 0;
+    double closest_value = array[0];
+    
+    for( int i=1; i < array_length; ++i )
+    {
+        if( std::abs(x - closest_value) > std::abs(x - array[i]) )
+        {
+            closest_value = array[i];
+            index = i;
+        }
+    }
+    
+    return index;
+}
```
Why do we need this overloaded version now? In `CalculateRadiativeLoss`, it looks like `find_closest` is only being called on vectors.

jwreep replied:
Mostly a placeholder while I was converting between the two.  Removed!

Commit:
remove overloaded find_closest

wtbarnes commented on source/loop.cpp:
```diff
@@ -255,6 +305,11 @@ void Loop::SaveResults(int i,double time)
     results.velocity[i] = velocity;
   }
+  
+  if( parameters.use_lookup_table_losses )
+  {
+      parameters.previous_density = __state[2];
+  }
```
I was surprised to see this in `SaveResults`. Is there a particular reason that it's here?

jwreep replied:
Good question.  Let me look around, because I can't remember right now why I put this there.

jwreep replied:
Might be a dumb question on my part, but where is the density calculated?  I'm storing the previous time-step's density here to later use for a weighted average.  I'm not seeing the actual calculation.  I'm guessing I placed it here because this is just a function for storing the calculated values.

wtbarnes replied:
Not a dumb question. The actual calculation of all the quantities (including density) is done internally by the boost integrator(s) and then stored in `state`. If you just always need the previous timestep's density, you can always grab it from `results.density[i-1]` if you know the current time index.

jwreep replied:
Right, got it.  The issue is then that the function that uses this, `CalculateAbundanceFactor`, doesn't have the time index.  So alternatively I could try to pass it the time index, which might be done by storing it in a similar fashion in parameters.  Or is there a clever way to get that index?

wtbarnes replied:
In that case, I wouldn't worry too much about it. Passing around the time index seems messier.

jwreep replied:
Sounds good.  In the future, if there's another function that needs the time index, then, perhaps it would be good to revisit this.

wtbarnes committed:
add tests for new radiative loss option

wtbarnes committed:
document radiative loss options

wtbarnes commented:
Ok, I've added a really basic test for each of the new radiative loss options to essentially just make sure that it can be run with that option without erroring. I'm not sure why you were seeing the error you were seeing with pytest.

Because I've pushed to your branch, make sure to pull down the latest changes before making any more commits to avoid any "fun" conflicts!

jwreep commented:
Thanks for the help with that!  I have no idea the cause of the error.  Pytest runs A-okay without that added code, though.

wtbarnes commented:
Thanks for all of your efforts per usual @jwreep!
